src/HtmlProcessor/HtmlParser.php

```diff
@@ -0,0 +1,208 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Pelago\Emogrifier\HtmlProcessor;
+
+/**
+ * Parse a HTML document fragment and add missing root elements.
+ *
+ * @author Kieran Brahney <kieran@supportpal.com>
+ */
+class HtmlParser
+{
+    /**
+     * Structure of a basic HTML document.
+     *
+     * @var array
+     */
+    protected $tree = [
+        'doctype' => '',
+        'html'    => [
+            'start'   => '<html>',
+            'end'     => '</html>',
+            'content' => [],
+        ],
+        'head'    => [
+            'start'   => '<head>',
+            'end'     => '</head>',
+            'content' => [],
+        ],
+        'body'    => [
+            'start'   => '<body>',
+            'end'     => '</body>',
+            'content' => [],
+        ],
+    ];
+
+    /**
+     * What root element did we last add to.
+     *
+     * @var string|null
+     */
+    protected $previousKey = null;
+
+    /**
+     * Parse a HTML document.
+     *
+     * @param  string  $html
+     *
+     * @return void
+     *
+     * @psalm-suppress MixedOperand
+     */
+    public function loadHtml(string $html)
+    {
+        $i = 0;
+        $len = \strlen($html);
+        while ($i < $len) {
+            if ($html[$i] == '<') {
```

Comment:
Should be `===`?  Also in a few other places.

Reply:
Yeah